File: 프로그래머스/lv2/68645. 삼각 달팽이/삼각 달팽이.py
import math
def solution(n):
    array = [0] * math.floor((n*(n+1)/2))
    turn_count,value,index,add_value = 0,0,0,1
    while(turn_count < n):
        if(turn_count % 3 == 0):
            for i in range(0,n - turn_count):
                index += value
                array[index] = add_value
                add_value += 1
                value += 1

        elif(turn_count % 3 == 1):
            for i in range(0,n - turn_count):
                index += 1
                array[index] = add_value
                add_value += 1

        else:
            for i in range(0, n-turn_count):
                index -= value
                array[index] = add_value
                add_value += 1
                value -= 1

        turn_count += 1
    return array

File: 프로그래머스/lv2/68645. 삼각 달팽이/삼각 달팽이.java
import java.util.*;

class Solution {
    static List<List<Integer>> direction = Arrays.asList(Arrays.asList(0,1), Arrays.asList(1,0), Arrays.asList(-1,-1)); 
    public int[] solution(int n) {
        int[][] matrix = new int[n][n];
		int[] array = new int[(n*(n+1))/2];
		
        int turnCount = 0;
        int index = 0;
        int addValue = 1;
        int x = 0, y = -1;
        
        while(true) {
        	int nextX = x + direction.get(index).get(0);
        	int nextY = y + direction.get(index).get(1);
        	
        	if(canGo(nextX, nextY, matrix) && matrix[nextY][nextX] == 0) {
        		turnCount = 0;
        		matrix[nextY][nextX] = addValue++;
        		x = nextX;
        		y = nextY;
        		
        	}else {
        		if(turnCount == 3) {
        			break;
        		}
        		turnCount++;
        		index = (index + 1) % 3;
        	}
        }
        
        index = 0;
        for(y = 0; y < matrix.length; y++) {
        	for(x = 0; x <= y; x++) {
        		array[index++] = matrix[y][x];
        	}
        }
        return array;
	}
	static boolean canGo(int x, int y, int[][] matrix) {
		return (0 <= x && x < matrix[0].length) && (0 <= y && y < matrix.length);
	}
}